Adding CVSS3.0 Score to the DefectDojo

flkhndlr opened the pull request:
Documentation: added the CVSS calculator v3.0 as a metric to findings. Field is optional. Through a click on the field the CVSS calculator drops down. Can create base, temporal and environmental score. Severity is set through JS as score changes.

**Note: DefectDojo is now on Python3 and Django 2.2.1 Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

When submitting a pull request, please make sure you have completed the following checklist:

- [x] Your code is flake8 compliant 
- [x ] Your code is python 3.5 compliant
- [x] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [x ] Model changes must include the necessary migrations in the dojo/dd_migrations folder.
- [x] Add applicable tests to the unit tests.

valentijnscholten commented:
Are there any existing parsers that already have this field available and should be updated to populate this new field in Defect Dojo?

flkhndlr commented:
> Are there any existing parsers that already have this field available and should be updated to populate this new field in Defect Dojo?

I'm really sorry, I'm not aware of any, but maybe there are. Maybe you have one in mind, that should be updated. Feel free to tell me.

valentijnscholten commented:
There are some cvss fields in the sample scan files, might be doable to modify those parsers to set the cvss field? If we have a cvss field, people are probably going to expect it is set whenever possible for the existing set of parsers.

See https://github.com/DefectDojo/sample-scan-files/search?q=cvss&unscoped_q=cvss

flkhndlr commented:
It was not my intend to modify these parsers. The scope of the pull request was just to implement the cvss score field. There are many different parsers and I think every maintainer of the respective parser should decide by themselves if they want to implement it.

valentijnscholten commented:
There are no maintainers per parser :-) But I can understand that it can be useful also to have this field in the UI/DB even if it is not (yet) populated by parsers. Can you rebase your branch, look at the copy-and-paste error I marked above? After that I'll let the other reviewers decide if they are happy to merge as i (without parsers updated) :-)

flkhndlr commented:
Hey, this is my first pull request, so I should rebase from the DefectDojo dev branch to my fork? And what copy-and-paste errors do you mean, I havent found them above.

valentijnscholten commented:
I think the PR now looks better indeed based of the dev branch, copy paste is this:
![image](https://user-images.githubusercontent.com/4426050/74922995-5904a100-53d0-11ea-8664-6596fb455351.png)

flkhndlr commented:
now I updated the repository and made some adjustments.

jvz commented:
Ooh this would be nice. Yes, there are a handful of security scanners that provide CVSS vectors (sometimes even the score, too). Currently, they're usually included in either the description or the severity justification depending on who wrote the parser.

As for that regex validation, I'd be careful not to be too strict. If you're only supporting 3.0, it might make sense to indicate that in the field name; otherwise, the regex could be loosened to allow for 2.0 vectors, 3.1 vectors, and whatever else eventually gets published.

This feature would mesh well with #1602 since CVSS vectors are typically included with every CVE, so that would allow for missing data to be filled in.

madchap commented:
Should not we hide the whole CVSS3 header in the finding's view if there is nothing to show?

madchap commented:
Should we seize that occasion to actually have the cvss string as you've done, and also the cvss integer score?

valentijnscholten commented:
There are new migrations in `dev`, could you please update this PR accordingly?

flkhndlr commented on dojo/static/dojo/js/cvss_calculator.js:
Labels are added through the cvss_helptext.js when page loads.

flkhndlr commented:
I updated the fork to the current state and renamed the cvss score to cvssv3. AccessLint issues are related to helptext which is reloaded and not hardcoded. I think further improvement should be part of a different PR.

madchap commented on dojo/models.py:
```diff
@@ -1427,6 +1429,7 @@ def long_desc(self):
         long_desc += '*' + self.title + '*\n\n'
         long_desc += '*Severity:* ' + str(self.severity) + '\n\n'
         long_desc += '*Cve:* ' + str(self.cve) + '\n\n'
+        long_desc += '*Cve:* ' + str(self.cvss) + '\n\n'
```
Apparently fixed:

`long_desc += '*CVSSv3.0:* ' + str(self.cvssv3) + '\n\n'`

flkhndlr commented:
is it normal, that there exists no Travis CI check?

valentijnscholten commented:
> is it normal, that there exists no Travis CI check?

travis is a bit weird sometimes, and sometimes we break it also. if you close and open the PR it will trigger Travis again.

flkhndlr commented:
Made the migration.

madchap commented:
I think the `dev` merge introduced some flake8 issues, per Travis.

valentijnscholten commented:
@flkhndlr any plans on finishing this one? even without the calculator it could be handy to have the cvss3 score available in the dojo data model.

flkhndlr commented:
Yeah I'm ready to finish this PR. I migrated it several times and I don't get, why the PR isn't approved. Is there something missing?

madchap commented:
I don't think so, apologies for that. It's going to get a bit of attention now to proceed. I had tested it locally, and I thought it was good :+1:

valentijnscholten commented:
Me neither, I just saw the latest comment:

> I think the dev merge introduced some flake8 issues, per Travis.

So maybe we all thought there was still some work today, but looks like latest Travis run was OK. But it has been a while and there are some conflicts now. Could you look into resolving those? I will try to get some attention for the PR on our slack channel to approvals / merging afterwards.

flkhndlr commented:
Ok I gonna update my branch and resolve the conflicts.

madchap commented:
I was thinking, would it be something to do to move to cvss 3.1 instead of 3.0 now?

flkhndlr commented:
Thats not really my intention. Im sorry

madchap commented:
That's cool, no problem. We can evolve it later when this one gets merged as is. Cheers!

flkhndlr commented:
Made the changes. So hope it's ready to merge

mtesauro reviewed (approved):
Approved.

flkhndlr commented:
Thank you
